File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/variables/types/SelectEditorFragment.java
package ch.rmy.android.http_shortcuts.variables.types;

import android.support.annotation.NonNull;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;

import butterknife.Bind;
import ch.rmy.android.http_shortcuts.R;
import ch.rmy.android.http_shortcuts.realm.models.Option;
import ch.rmy.android.http_shortcuts.realm.models.Variable;

public class SelectEditorFragment extends VariableEditorFragment {

    @Bind(R.id.select_options_list)
    LinearLayout optionsList;
    @Bind(R.id.select_options_add_button)
    Button addButton;

    private Variable variable;

    @Override
    protected int getLayoutResource() {
        return R.layout.variable_editor_select;
    }

    @Override
    protected void setupViews(View parent) {
        addButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                showEditDialog(null, -1);
            }
        });
    }

    @Override
    public void updateViews(Variable variable) {
        this.variable = variable;
        optionsList.removeAllViews();
        int i = 0;
        for (Option option : variable.getOptions()) {
            optionsList.addView(createOptionView(option, i));
            i++;
        }
    }

    private View createOptionView(final Option option, final int index) {
        View optionView = getLayoutInflater(null).inflate(R.layout.select_option, optionsList, false);
        ((TextView) optionView.findViewById(R.id.select_option_label)).setText(option.getLabel());
        optionView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                showEditDialog(option, index);
            }
        });
        return optionView;
    }

    private void showEditDialog(final Option option, final int index) {
        View editorView = getLayoutInflater(null).inflate(R.layout.select_option_editor_item, null);
        final TextView labelInput = (TextView) editorView.findViewById(R.id.select_option_label);
        final TextView valueInput = (TextView) editorView.findViewById(R.id.select_option_value);

        if (option != null) {
            labelInput.setText(option.getLabel());
            valueInput.setText(option.getValue());
        }
        MaterialDialog.Builder builder = new MaterialDialog.Builder(getContext())
                .title(R.string.title_add_select_option)
                .customView(editorView, true)
                .positiveText(R.string.dialog_ok)
                .onPositive(new MaterialDialog.SingleButtonCallback() {
                    @Override
                    public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                        String label = labelInput.getText().toString();
                        String value = valueInput.getText().toString();
                        if (option != null) {
                            updateOption(option, label, value);
                        } else {
                            addNewOption(label, value);
                        }
                    }
                })
                .negativeText(R.string.dialog_cancel);

        if (option != null) {
            builder = builder
                    .neutralText(R.string.dialog_remove)
                    .onNeutral(new MaterialDialog.SingleButtonCallback() {
                        @Override
                        public void onClick(MaterialDialog dialog, DialogAction which) {
                            removeOption(index);
                        }
                    });
        }

        builder.show();
    }

    private void addNewOption(String label, String value) {
        Option option = Option.createNew(label, value);
        variable.getOptions().add(option);
        updateViews(variable);
    }

    private void updateOption(Option option, String label, String value) {
        option.setLabel(label);
        option.setValue(value);
        updateViews(variable);
    }

    private void removeOption(int index) {
        variable.getOptions().remove(index);
        updateViews(variable);
    }

    @Override
    public boolean validate() {
        if (variable.getOptions().isEmpty()) {
            new MaterialDialog.Builder(getContext())
                    .content(R.string.error_not_enough_select_values)
                    .positiveText(R.string.dialog_ok)
                    .show();
            return false;
        }
        return true;
    }

}

File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/variables/types/SelectEditorFragment.kt
package ch.rmy.android.http_shortcuts.variables.types

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.widget.Button
import android.widget.EditText
import ch.rmy.android.http_shortcuts.R
import ch.rmy.android.http_shortcuts.adapters.SelectVariableOptionsAdapter
import ch.rmy.android.http_shortcuts.realm.models.Option
import ch.rmy.android.http_shortcuts.realm.models.Variable
import ch.rmy.android.http_shortcuts.utils.Destroyer
import ch.rmy.android.http_shortcuts.utils.DragOrderingHelper
import ch.rmy.android.http_shortcuts.utils.mapIf
import ch.rmy.android.http_shortcuts.utils.showIfPossible
import ch.rmy.android.http_shortcuts.utils.showMessageDialog
import ch.rmy.android.http_shortcuts.variables.VariableButton
import ch.rmy.android.http_shortcuts.variables.VariableEditText
import com.afollestad.materialdialogs.MaterialDialog
import kotterknife.bindView

class SelectEditorFragment : VariableEditorFragment() {

    private var variable: Variable? = null

    override val layoutResource = R.layout.variable_editor_select

    private val selectOptionsAddButton: Button by bindView(R.id.select_options_add_button)
    private val selectOptionsList: RecyclerView by bindView(R.id.select_options_list)
    private val optionsAdapter = SelectVariableOptionsAdapter()

    override fun setupViews() {
        selectOptionsAddButton.setOnClickListener { showAddDialog() }
        selectOptionsList.layoutManager = LinearLayoutManager(context)
        selectOptionsList.adapter = optionsAdapter
        optionsAdapter.clickListener = this::showEditDialog
        initDragOrdering()
    }

    private fun initDragOrdering() {
        val dragOrderingHelper = DragOrderingHelper()
        dragOrderingHelper.positionChangeSource.add { (oldPosition, newPosition) ->
            variable!!.options!!.move(oldPosition, newPosition)

            optionsAdapter.notifyItemMoved(oldPosition, newPosition)
        }.attachTo(destroyer)
        dragOrderingHelper.attachTo(selectOptionsList)
    }

    override fun updateViews(variable: Variable) {
        this.variable = variable
        optionsAdapter.options = variable.options!!
        optionsAdapter.notifyDataSetChanged()
    }

    private fun showAddDialog() {
        showEditDialog(null)
    }

    private fun showEditDialog(option: Option?) {
        val destroyer = Destroyer()

        val editorView = layoutInflater.inflate(R.layout.select_option_editor_item, null)
        val labelInput = editorView.findViewById<EditText>(R.id.select_option_label)
        val valueInput = editorView.findViewById<VariableEditText>(R.id.select_option_value)
        val valueVariableButton = editorView.findViewById<VariableButton>(R.id.variable_button_value)

        valueInput.bind(valueVariableButton, variableKeyProvider).attachTo(destroyer)

        if (option != null) {
            labelInput.setText(option.label)
            valueInput.rawString = option.value
        }

        MaterialDialog.Builder(context!!)
                .title(if (option != null) R.string.title_edit_select_option else R.string.title_add_select_option)
                .customView(editorView, true)
                .positiveText(R.string.dialog_ok)
                .onPositive { _, _ ->
                    val label = labelInput.text.toString()
                    val value = valueInput.rawString
                    if (option != null) {
                        updateOption(option, label, value)
                    } else {
                        addNewOption(label, value)
                    }
                }
                .negativeText(R.string.dialog_cancel)
                .mapIf(option != null) {
                    it
                            .neutralText(R.string.dialog_remove)
                            .onNeutral { _, _ -> removeOption(option!!) }
                }
                .dismissListener {
                    destroyer.destroy()
                }
                .showIfPossible()
    }

    private fun addNewOption(label: String, value: String) {
        val option = Option.createNew(label, value)
        variable!!.options!!.add(option)
        updateViews(variable!!)
    }

    private fun updateOption(option: Option, label: String, value: String) {
        option.label = label
        option.value = value
        updateViews(variable!!)
    }

    private fun removeOption(option: Option) {
        variable!!.options!!.removeAll { it.id == option.id }
        updateViews(variable!!)
    }

    override fun validate() = if (variable!!.options!!.isEmpty()) {
        showMessageDialog(R.string.error_not_enough_select_values)
        false
    } else {
        true
    }

}
